server-core/service/src/main/java/io/infectnet/server/service/impl/UserServiceImpl.java

```diff
@@ -1,22 +1,105 @@
+     *      must contain at least one number
+     *      must contain at least one uppercase letter.
+     * If authentication fails no new user will be stored, otherwise a new User is stored
+     * with the username, the email, the password and the date of the registration now produced.
+     *
+     * @param token the token from the current registration
+     * @param email the email address from the current registration
+     * @param username the username from the current registration
+     * @param password the password from the current registration
+     * @return the Optional containing the UserDTO made from the created user
+     */
     @Override
-    public UserDTO register(String token, String username, String password) {
-        return null;
+    public Optional<UserDTO> register(String token, String email, String username, String password) {
+        if(tokenIsValid(token) && emailIsUnique(email) && usernameIsUnique(username) && passwordIsValid(password)){
+            UserDTO newUser = new UserDTO(username,email,password, LocalDateTime.now());
+            User user = converterService.map(newUser, User.class);
+            userStorage.saveUser(user);
+            return Optional.of(newUser);
+        }else{
+            return Optional.empty();
+        }
     }
 
     @Override
-    public UserDTO login(String username, String password) {
-        return null;
+    public Optional<UserDTO> login(String username, String password) {
+        Optional<User> user = userStorage.getUserByUserName(username);
+        if(user.isPresent() && user.get().getPassword().equals(password)){
+            UserDTO userDTO = converterService.map(user.get(),UserDTO.class);
+            return Optional.of(userDTO);
+        }else{
+            return Optional.empty();
+        }
     }
 
     @Override
     public boolean exists(UserDTO user) {
-        return false;
+        return userStorage.exists(converterService.map(Objects.requireNonNull(user), User.class));
+
+    }
+
+    /**
+     * Checks if the token given is in the storage and not yet expired.
+     *
+     * @param token the token to check
+     * @return true if the token is stored and valid, otherwise false
+     */
+    private boolean tokenIsValid(String token) {
+        Optional<Token> foundToken = tokenStorage.getTokenByTokenString(token);
+        return foundToken.isPresent();
+                //&& foundToken.get().getExpirationDateTime().until(LocalDateTime.now(), ChronoUnit.MINUTES)<=tokenService.getExpirationLimit());
+    }
+
+    private boolean emailIsUnique(String email) {
+        Optional<User> user = userStorage.getUserByEmail(email);
+        return !user.isPresent();
+    }
+
+    private boolean usernameIsUnique(String username) {
+        Optional<User> user = userStorage.getUserByUserName(username);
+        return !user.isPresent();
+    }
+
+    private boolean passwordIsValid(String password) {
+        return (password.length()>=8
```

Comment:
Please extract these values into named constants. The regular expressions should be precompiled `Pattern`s.